src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp

```diff
@@ -44,30 +44,74 @@ struct Variables;
+// two signatures of callback functions.
+// Uses SFINAE on return type.
+template <typename T, typename DbTags, typename Metavariables>
+auto apply_callback(
+    const gsl::not_null<db::DataBox<DbTags>*> box,
+    const gsl::not_null<Parallel::ConstGlobalCache<Metavariables>*> cache,
+    const typename Metavariables::temporal_id& temporal_id) noexcept
+    -> decltype(T::post_interpolation_callback::apply(box, cache, temporal_id),
+                bool()) {
+  return T::post_interpolation_callback::apply(box, cache, temporal_id);
+}
+
+template <typename T, typename DbTags, typename Metavariables>
+auto apply_callback(
+    const gsl::not_null<db::DataBox<DbTags>*> box,
+    const gsl::not_null<Parallel::ConstGlobalCache<Metavariables>*> cache,
+    const typename Metavariables::temporal_id& temporal_id) noexcept
+    -> decltype(T::post_interpolation_callback::apply(*box, *cache,
+                                                      temporal_id),
+                bool()) {
+  T::post_interpolation_callback::apply(*box, *cache, temporal_id);
+  // For the simpler callback function, we will always clean up volume data, so
+  // we return true here.
+  return true;
+}
+
 /// Calls the callback function, tells interpolators to clean up the current
 /// temporal_id, and then if there are more temporal_ids to be interpolated,
 /// starts the next one.
-template <typename InterpolationTargetTag,
-          typename DbTags, typename Metavariables>
+template <typename InterpolationTargetTag, typename DbTags,
+          typename Metavariables>
 void callback_and_cleanup(
     const gsl::not_null<db::DataBox<DbTags>*> box,
     const gsl::not_null<Parallel::ConstGlobalCache<Metavariables>*>
         cache) noexcept {
   const auto temporal_id =
       db::get<Tags::TemporalIds<Metavariables>>(*box).front();
 
-  // Do the callback
-  const auto& new_box = db::create_from<
+  // Add compute_items_on_target to the box.
+  auto box_with_more_items = db::create_from<
       db::RemoveTags<>, db::AddSimpleTags<>,
       db::AddComputeTags<
-          typename InterpolationTargetTag::compute_items_on_target>>(*box);
-  InterpolationTargetTag::post_interpolation_callback::apply(new_box, *cache,
-                                                             temporal_id);
-
-  // In the future, when generalizing this function to accommodate
-  // horizon finding, we will need to think more carefully about
-  // control flow and the point at which the temporal_id should be
-  // popped from the list.
+          typename InterpolationTargetTag::compute_items_on_target>>(
+      std::move(*box));
+
+  // apply_callback should return true if we are done with this
+  // temporal_id.  It should return false only if the callback
+  // calls another `intrp::Action` that needs the volume data at this
+  // same temporal_id.  If it returns false, we exit here and do not
+  // clean up.
+  const bool done_with_temporal_id = apply_callback<InterpolationTargetTag>(
```

Comment:
This addition you can test by extending existing tests.

Reply:
Modified Test_InterpolationTargetReceiveVars to test this.